### DISCUSSION TITLE
How to call PowerShell script from C# and pass data via standard input to it

### DISCUSSION THREAD

**Discussion by michaelwildvarian** [PowerShell/PowerShell#24616] [Q&A]
Environment: Windows 11 23H2, PowerShell 7.4.6

Hi all, I'd like to be able to extend my C# program with "hook" scripts. These scripts should receive their input from stdin and output their results to stdout. Ideally I'd like these scripts to be in PowerShell due to the cross-platform support.

What I got so far (simplified, with no error handling etc., to illustrate the problem):

```cs
// Program.cs
using System.Diagnostics;

ProcessStartInfo startInfo = new()
{
  UseShellExecute = false,
  FileName = "pwsh.exe",
  Arguments = $"-NoProfile -File ./someHook.ps1",
  RedirectStandardInput = true,
  RedirectStandardOutput = true,
  CreateNoWindow = true,
};

using Process process = new();
process.StartInfo = startInfo;
process.EnableRaisingEvents = true;

process.Start();
await using var stdinWriter = process.StandardInput;
if (stdinWriter.BaseStream.CanWrite)
{
  await stdinWriter.WriteAsync("SOME INPUT\n\n");
  stdinWriter.Close();
}

await process.WaitForExitAsync();

var output = await process.StandardOutput.ReadToEndAsync();

Console.Write(output);
```

The script looks like this:

```ps1
# someHook.ps1
param(
  [Parameter(
    Position=0,
    Mandatory=$true,
    ValueFromPipeline=$true,
    ValueFromPipelineByPropertyName=$true
  )]
  [String]$InputData
)

begin {
  echo "START>>>"
}

process {
  echo "The input was:"
  echo $InputData
}

end {
  echo "<<<<END"
}
```

When I run this, the script is invoked all right and it can read from the pipeline. However, the standard output always contains the data from standard input prepended:

```
> dotnet run
SOME INPUT
START>>>
The input was:
SOME INPUT
<<<<END
```

This does not happen if the hook script is a Python script instead, or a simple C++ executable.

Is there anything wrong with how I call the PowerShell script? Or am I being hit by #3223?

**Comment by rhubarb-geek-nz** [comment]
Hi, PowerShell is not classic stdin/stdout friendly. It has it's own pipelines.

The first option is to run your PowerShell scripts directly in the C# application using the PowerShell SDK. Then use [PowerShell.Invoke](https://learn.microsoft.com/en-us/dotnet/api/system.management.automation.powershell.invoke?view=powershellsdk-7.4.0) with your pipeline data.

The second is to use a cut down [PowerShell executable](https://github.com/rhubarb-geek-nz/powershell-executive) that runs the script without the existing stdin/stdout wrapping and lets the script do exactly what it wants.

An additional module is [rhubarb-geek-nz/Console](https://github.com/rhubarb-geek-nz/Console) which wraps the console read and write functions with the PowerShell pipelines.

**Reply by michaelwildvarian** [reply]
Thanks. That's a "NO" to PowerShell then. I am not going to compromise the flexibility of being able to write the hooks in any possible language for this niche language that decides to do things differently just for the sake of messing with people...

**Reply by rhubarb-geek-nz** [reply]
> Thanks. That's a "NO" to PowerShell then.

That is a shame, given you are already using C# then integrating with the PowerShell SDK would allow the scripts to execute in your process and directly access internal objects.

Also it somebody chooses to use PowerShell in a consistent way with stdin and stdout as a hook, that would be their choice. If you define the contract as being a command line along with stdin/stdout it is up to the hook implementer how they meet that.
